Tool refactor, multi tools

Zumorica opened the pull request:
- Refactors all tools (except Welder) into a single component, ToolComponent
- ~~Adds a bunch of component interfaces such as IWrenchAct, ICrowbarAct...~~
- Adds MultiToolComponent, which can be used to make tools that can switch behaviors (See jaws of life, power drill) 
- Adds welding fuel chemical
- Construction now uses Tool enum.
- Renames sound_collection to soundCollection.
- Renames Wrenchable to Anchorable. ~~Refactors it to use IWrenchAct.~~

Next up is moving all components to use the new component interfaces for tools. But that can be done by someone else.

RemieRichards commented on Content.Server/GameObjects/EntitySystems/Click/InteractionSystem.cs:
```diff
@@ -43,6 +46,129 @@ public class AttackByEventArgs : EventArgs
+    public interface IToolAct
+    {
+        /// <summary>
+        /// Called when using a wrench on an entity
```
are you sure about that?

Zumorica replied:
And this is why you double check when copy-pasting code around--

RemieRichards commented on Content.Server/GameObjects/EntitySystems/Click/InteractionSystem.cs:
```diff
@@ -43,6 +46,129 @@ public class AttackByEventArgs : EventArgs
+        /// <summary>
+        /// Called when using a wrench on an entity
+        /// </summary>
+        bool ToolAct(ToolActEventArgs eventArgs) => false;
```
This method isn't used, is it?

Zumorica replied:
Ah, I forgot to actually call it! I fixed that now

Hugal31 commented on Content.Server/GameObjects/Components/Construction/ConstructionComponent.cs:
```diff
@@ -141,46 +142,42 @@ bool TryProcessStep(ConstructionStep step, IEntity slapped)
                 case ConstructionStepTool toolStep:
+                    if (!slapped.TryGetComponent<ToolComponent>(out var tool))
+                        return false;
                     switch (toolStep.Tool)
```
Here, you do a lot of `if (tool.Behavior ==`. You could factor it out and do `if (tool.Behavior == toolStep.Tool)` before the `switch`.

This way, you could factor some cases, like:
```c#
case Tool.CrowBar:
case Tool.Wrench:
case Tool.Wirecutters:
    tool.PlayUseSound();
    return true;
```

Alternatively, if you add TryCrowBar, TryWrench, TryWireCutters, ... and check if the current Behavior is the right tool in this methods, (like you already to in TryWeld, for instance), you can just do:
```c#
switch (toolStep.Tool)
{
case ToolType.CrowBar:
    if (tool.TryCrowBar())
    {
        tool.PlayUseSound();
        return true;
    }
    return false;
// Repeat for all tools, whithout need to check for tool.Behavior
}

Zumorica replied:
You're right, I could have cleaned up that code. I only made it compatible with the tool refactor without taking into account cleanup opportunities like these.
I cleaned it up now.

> Alternatively, if you add TryCrowBar, TryWrench, TryWireCutters, ... and check if the current Behavior is the right tool in this methods, (like you already to in TryWeld, for instance), you can just do [...]

I'm against that, TryWeld is only used to check if the welder has enough fuel to weld something. For the rest of these, you can just compare the Behavior like on your first suggestion, which I ended up doing.

Hugal31 commented on Content.Server/GameObjects/Components/Construction/ConstructionComponent.cs:
```diff
@@ -144,46 +144,10 @@ bool TryProcessStep(ConstructionStep step, IEntity slapped)
-                    }
+                    if (toolStep.Tool != tool.Behavior) return false;
+                    tool.PlayUseSound();
+                    return true;
```
You removed `tool.TryWeld(toolStep.Amount)`, is this on purpose?

Zumorica replied:
Not at all, my bad. Fixed now.

Zumorica commented:
Alright, I removed all the welder-specific stuff and moved it to its own component.
I think creating custom tools should be trivial if your component inherit ToolComponent.
This is ready for review once again.

Hugal31 commented on Content.Server/GameObjects/Components/Doors/AirlockComponent.cs:
```diff
@@ -200,27 +201,28 @@ private bool IsPowered()
-                {
-                    Close();
-                }
+            if (tool.Behavior != Tool.Crowbar) return false;
```
I see you are alternating one liners and two liners `if`s and `[ViewVariable] public bool MyVar;`
I find this confusing.

Zumorica replied:
Alright, I cleaned up the code a bit to address most of this. If you find any more examples of this, leave a review.

RemieRichards commented on Content.Client/GameObjects/Components/Interactable/WelderComponent.cs:
```diff
@@ -0,0 +1,89 @@
+        [ViewVariables] public float Fuel { get; private set; }
+        [ViewVariables] public bool Activated { get; private set; }
+        [ViewVariables]
+        public override Tool Behavior
```
```csharp
public override Tool Behavior => _behaviour;
```

Zumorica replied:
Dddone

RemieRichards commented on Content.Server/GameObjects/Components/Interactable/ToolComponent.cs:
```diff
@@ -122,18 +88,24 @@ public override void Initialize()
+                }
+                catch
+                {
+                    // ignored
```
Don't we have somewhere you could log this.

Zumorica replied:
Thing is, inherited components like the Welder might not use the behavior field... So I'm not sure if we should really log that hmm

Zumorica replied:
I changed this now to not rely on a try-catch block

PJB3005 commented on Content.Server/GameObjects/Components/Construction/ConstructionComponent.cs:
```diff
@@ -141,52 +142,13 @@ bool TryProcessStep(ConstructionStep step, IEntity slapped)
+                    if (!slapped.TryGetComponent<ToolComponent>(out var tool))
+                        return false;
+                    if (toolStep.Tool != tool.Behavior) return false;
+                    if (toolStep.Tool == Tool.Welder && !((WelderComponent)tool).TryWeld(toolStep.Amount)) return false;
```
So the problem here is that the powered hand drill you added now can't actually be, well, powered, because you hardcoded it so welders are special cased.

Zumorica replied:
I changed it so a virtual UseTool is called on the ToolComponent, making it possible for actual powered tools to exist.
I called Welder's own overload here due to toolsteps specifying the amount of fuel needed.

PJB3005 commented on Content.Server/GameObjects/Components/Doors/AirlockComponent.cs:
```diff
@@ -200,27 +201,25 @@ private bool IsPowered()
 
         public bool AttackBy(AttackByEventArgs eventArgs)
```
Why isn't this crowbar act now?

PJB3005 commented on Content.Server/GameObjects/Components/Gravity/GravityGeneratorComponent.cs:
```diff
@@ -98,29 +98,25 @@ bool IAttackHand.AttackHand(AttackHandEventArgs eventArgs)
-                damagable.HealAllDamage();
-                breakable.broken = false;
-                _intact = true;
+            var welder = (WelderComponent)eventArgs.ToolComponent;
```
If there should always be a `WelderComponent` doing the welding then the `eventArgs` should expose that type directly. If this is not the case then this code is buggy.

Zumorica replied:
This isn't an issue anymore.

PJB3005 commented on Content.Client/GameObjects/Components/Interactable/ToolComponent.cs:
```diff
@@ -0,0 +1,75 @@
+        private bool _statusShowBehavior;
+
+        [ViewVariables(VVAccess.ReadWrite)] private bool _uiUpdateNeeded;
+        [ViewVariables] public bool StatusShowBehavior => _statusShowBehavior;
```
This `StatusShowBehavior` thing should absolutely be a part of multitool, not the base `ToolComponent`.

Zumorica replied:
I just removed client-side ToolComponent in favor of client-side MultiToolComponent

PJB3005 commented on Content.Server/GameObjects/Components/Interactable/ToolComponent.cs:
```diff
@@ -0,0 +1,182 @@
+
+        public void AfterAttack(AfterAttackEventArgs eventArgs)
+        {
+            if (Behavior != Tool.Crowbar)
```
I *really* don't like this being hardcoded into the base `ToolComponent` like this.

At the same time I also recognize why it is the way it is, hrm.

ComicIronic replied:
I don't see why it is - could you explain? Is a `CrowbarActAfter` not somehow appropriate?

PJB3005 replied:
I guess that might work.

Zumorica replied:
Hmm, I don't see where I should put this logic. Any ideas?

Zumorica replied:
In the end I made this its own component.

PJB3005 commented on Content.Server/GameObjects/Components/Power/PowerTransferComponent.cs:
```diff
@@ -140,17 +141,16 @@ public bool CanConnectTo()
 
         public bool AttackBy(AttackByEventArgs eventArgs)
```
Any reason this isn't using `IWirecutterAct`?

Zumorica replied:
I removed ToolActs now.

PJB3005 commented on Content.Server/GameObjects/EntitySystems/Click/InteractionSystem.cs:
```diff
@@ -561,9 +676,98 @@ public void Interaction(IEntity user, IEntity weapon, IEntity attacked, GridCoor
+            }
+
+            // We handle specific tools AttackBy here.
+            if (weapon.TryGetComponent<ToolComponent>(out var tool))
```
This should absolutely not be inside `InteractionSystem`.

Zumorica replied:
Done.

Zumorica commented:
This is ready for review _again_.

Hugal31 commented on Content.Server/GameObjects/Components/Interactable/ToolComponent.cs:
```diff
@@ -0,0 +1,180 @@
+        {
+            get => _useSoundCollection;
+            set => _useSoundCollection = value;
+        }
```
Couldn't you use auto-properties for those?

Zumorica replied:
Nope. Serializer can't use auto-properties for DataField()

PJB3005 replied:
You actually can with the expression support that Q added now. `serializer.DataField(this, x => x.Foo, "foo", 1);`

Hugal31 commented on Content.Server/GameObjects/Components/Interactable/ToolComponent.cs:
```diff
@@ -0,0 +1,180 @@
+
+        public void AfterAttack(AfterAttackEventArgs eventArgs)
+        {
+            if (Qualities != ToolQuality.Prying)
```
I did not tested your branch, nor I am familiar with the AfterAttack system, but if you are assembling/disassembling something with the crowbar, wouldn't that still trigger this function and try to remove the tile?

Also, for readability, I would have create a function for prying tiles:
```c#
if (Qualities == ToolQuality.Prying)
    TryPryTile(eventArg);
}
```

Zumorica replied:
From the [component interfaces wiki page](https://hackmd.io/@ss14/component-interfaces):
```IAfterAttack
This interface gives components a behavior when clicking on another object and no interaction occurs, or when interacting with an entity out of range that has no behavior. Doesn’t pass what you clicked on as an argument.```

Also that's a good suggestion for readability.

Clyybber commented on Content.Server/GameObjects/Components/AnchorableComponent.cs:
```diff
@@ -1,5 +1,6 @@
-using Content.Server.GameObjects.Components.Interactable.Tools;
+using Content.Server.GameObjects.Components.Interactable;
 using Content.Server.GameObjects.EntitySystems;
+using Content.Shared.GameObjects.Components.Interactable;
```
```suggestion
```

Zumorica commented on Content.Server/GameObjects/Components/Interactable/MultitoolComponent.cs:
```diff
@@ -0,0 +1,124 @@
+
+            public void ExposeData(ObjectSerializer serializer)
+            {
+                if(serializer.Reading)
```
Done.

Zumorica commented on Content.Server/GameObjects/Components/Gravity/GravityGeneratorComponent.cs:
```diff
@@ -100,27 +101,26 @@ bool IAttackHand.AttackHand(AttackHandEventArgs eventArgs)
 
-                var entitySystemManager = IoCManager.Resolve<IEntitySystemManager>();
-                var notifyManager = IoCManager.Resolve<IServerNotifyManager>();
+            if (!tool.UseTool(eventArgs.User, Owner, ToolQuality.Welding, 5f))
```
> it would be neat if this showed a notification like "your welder is empty/off" if applicable.

I just added this in WelderComponent, making it show popups explaining why you couldn't use it. 

> `UseTool` doesn't distinguish between "this tool doesn't have this quality" and "this tool has this quality but is empty/off". Should it?

I personally think it shouldn't. The tool should handle this, showing a popup/notification to the user like you suggested and I added.

> I would expect in many cases for `AttackBy` to be handled when using an empty welder but then there's edge cases like closets where you probably don't (want to put welder into closet instead of deconstructing closet).

Hmmm, I'm not too sure what would be the best way to handle that.
